### PR TITLE
Add predicate support to accumulation checker

### PR DESCRIPTION
**Pull request by kelloggm**
Differences from predicate support in the object construction checker:
* sensible LUB/GLB between predicate annotations
* subtyping between predicates does not use implication (because I didn't want a dependency on an SMT solver), so it conservatively underapproximates
* unicode support in identifiers
* no dependency on an external parsing library other than JavaParser, which the CF already depends on

### CODE REVIEW

**Comment by kelloggm** [comment]
(sorry for the messy commit history)

File: framework/tests/accumulation/SimpleInferenceMerge.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import testaccumulation.qual.*;
+
+class SimpleInferenceMerge {
```
Comment (msridhar) [inline framework/tests/accumulation/SimpleInferenceMerge.java:3]: This test case is unrelated to predicates, right?

File: docs/manual/accumulation-checker.tex [hunk 2]
```diff
@@ -55,15 +55,21 @@
 \href{https://github.com/typetools/checker-framework/blob/master/framework/src/test/java/testaccumulation/qual/TestAccumulationBottom.java}{TestAccumulationBottom.java}.
 It should take no arguments, and should be a subtype of the accumulator type you defined earlier.
 
+You may also choose to define a predicate annotation, analogous to
+\href{https://github.com/typetools/checker-framework/blob/master/framework/src/test/java/testaccumulation/qual/TestAccumulationPredicate.java}{TestAccumulationPredicate.java}.
+A predicate annotation allows you to write specifications with disjunctions (the accumulation annotation implicitly already permits
```
Comment (msridhar) [inline docs/manual/accumulation-checker.tex:60]: We probably want some documentation of the unsound handling of negation

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationVisitor.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    /** Checks each predicate annotation to make sure the predicate is well-formed. */
+    @Override
+    public Void visitAnnotation(final AnnotationTree node, final Void p) {
+        AnnotationMirror anno = TreeUtils.annotationFromAnnotationTree(node);
+        if (atypeFactory.isPredicate(anno)) {
+            String errorMessage = atypeFactory.isValidPredicate(anno);
+            if (errorMessage != null) {
+                checker.report(
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationVisitor.java:33]: Use `reportError` rather than `report`.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationVisitor.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                        node,
+                        new DiagMessage(Diagnostic.Kind.ERROR, "predicate.invalid", errorMessage));
+                return null;
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationVisitor.java:36]: Why not give the superclass the opportunity to issue more errors, if there are others?

File: framework/tests/accumulation/SimpleInferenceMerge.java (hunk 1 above)
Comment (kelloggm) [reply framework/tests/accumulation/SimpleInferenceMerge.java:3]: Yeah. It probably should have been included in the original AccumulationChecker PR, but was left off somehow. I included it here since it passes and I moved over a bunch of the other tests, so it didn't seem like it needed its own.

**Comment by kelloggm** [comment]
@mernst I have addressed or responded to all of your comments. Thanks for being patient.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java [hunk 5]
```diff
@@ -58,38 +73,70 @@
...
+                rejectMalformedPredicate("have an element of type String");
+            }
+        }
+
         this.bottom = AnnotationBuilder.fromClass(elements, bottom);
         this.top = createAccumulatorAnnotation(Collections.emptyList());
 
-        // Every subclass must call postInit!  This does not do so for subclasses.
-        if (this.getClass() == AccumulationAnnotatedTypeFactory.class) {
-            this.postInit();
```
Comment (msridhar) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:91]: Wall calling `postInit()` here before a bug?

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -248,6 +320,21 @@ public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
...
+     *
+     * Predicate subtyping is defined as follows:
+     *
+     * <ul>
+     *   <li>An accumulator is a subtype of a predicate if substitution from the accumulator to the
+     *       predicate makes the predicate true. For example, {@code Acc(A)} is a subtype of {@code
+     *       AccPred("A || B")}, because when A is replaced with {@code true} and B is replaced with
+     *       {@code false}, the resulting boolean formula evaluates to true.
+     *   <li>A predicate P is a subtype of an accumulator iff after converting the accumulator into
+     *       a predicate representing the conjunction of its elements, P is a subtype of that
```
Comment (msridhar) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:332]: Does this mean that after converting the accumulator to a predicate, `P` must _equal_ that predicate (i.e., it falls into the next case of predicate subtyping)?

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java [hunk 7]
```diff
@@ -248,6 +320,21 @@ public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
      *        |
      *      bottom
      * </pre>
+     *
+     * Predicate subtyping is defined as follows:
```
Comment (msridhar) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:324]: It might be good to document what the "right thing" to do would be, and why we deviate from that.  I think the "right thing" here is logical implication, i.e., P is a subtype of Q iff P implies Q.  But we don't bake in the implication check to keep things simple and minimize dependencies.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java (hunk 5 above)
Comment (kelloggm) [reply framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:91]: The condition was always false, because this is an abstract class. All subclasses have to call `postInit`.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java (hunk 6 above)
Comment (kelloggm) [reply framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:332]: In the default implementation, yes, *but* if the accumulation checker writer has implemented a more complex predicate subtyping rule that would also be taken into account.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java (hunk 7 above)
Comment (kelloggm) [reply framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java:324]: I'll add a comment.
